torch/distributed/fsdp/flat_param.py

```diff
@@ -1413,7 +1433,14 @@ def _use_unsharded_grad_views(self) -> None:
             )  # did not save prefixed name
             param = getattr(module, param_name)
             prim_param = getattr(prim_module, prim_param_name)
-            param.grad = prim_param.grad
+            if param.shape != prim_param.grad.shape:
```

Comment:
just for my own education, where can I learn more about why do we need to set grad in the `if` clause. I assume we are not going to accumulate grad to sharded param gradients?

Reply:
For this `if`, the "primary parameter" `prim_param` must have `.grad` set from the previous `for` loop in this method.

Now, this `for` loop is over the shared parameters, and this `if` is just checking the same shape mismatch case as before (i.e. `param` is sharded but its gradient is unsharded).

We should still set the `param.grad` to be `prim_param.grad`, which is a view into the unsharded gradient because for `no_sync()`, we may accumulate an _unsharded_ gradient.

The invariant of `use_orig_params=True` path is that each original parameter and its gradient should always be a view into the underlying `flat_param` and `flat_param.grad`, respectively. Whether the user actually cares that we set `param.grad` to be an unsharded view into `flat_param.grad` during `no_sync()` is debatable, but this is an important design choice that makes reasoning about the system more tractable (at least to me). This is why I was a big proponent of the `_use_sharded_views`, `_use_sharded_grad_views`, `_use_unsharded_views`, `_use_unsharded_grad_views` approach, where any change to the `flat_param` storage should include calls to these methods.